### PR TITLE
feat: Histograms API & Elasticsearch Implementation

### LINKED ISSUE #1600: Implement Elasticsearch API
Implement the histogram API and the elasticsearch plugin logic in accordance with #1592 

** Might be wise to split into two PRs: ES impl and then the http endpoint

### PR DESCRIPTION
**Pull request by Tavh**
## What this PR does:
- Implements the "histograms" API defined in #1592
- Implements the Elasticsearch implementation for the histograms API
- Tests the histograms API

## Invocation examples:

**errors_distinct_count**:

`curl --location --request POST 'http://localhost:8080/v1/histograms' \
--header 'Content-Type: application/json' \
--data-raw '{
    "timeframe": {
        "startTimeUnixNanoSec": 0,
        "endTimeUnixNanoSec": 0
    },
    "filters": [
        {
            "keyValueFilter": {
                "key": "span.status.code",
                "operator":"in",
                "value": ["Error"]
            }
        }
    ],
    "interval": 600000,
    "intervalKey": "span.startTimeUnixNano",
    "aggregations": {
        "errors_distinct_count": {
            "func": "DISTINCT_COUNT",
            "groupBy": "span.attributes.http.status_code",
            "maxGroups": 5
        }    
    }
}'`

Screenshot 2023-05-17 at 13 33 07

**latency_percentiles**:

`curl --location --request POST 'http://localhost:8080/v1/histograms' \
--header 'Content-Type: application/json' \
--data-raw '{
    "timeframe": {
        "startTimeUnixNanoSec": 0,
        "endTimeUnixNanoSec": 0
    },
    "interval": 600000,
    "intervalKey": "span.startTimeUnixNano",
    "aggregations": {
        "latency_percentiles": {
            "func": "PERCENTILES",
            "key": "externalFields.durationNano",
            "aggregationParameters": {
                "percentiles": [50, 75, 90, 95, 99]
            }
        }    
    }
}'`

Screenshot 2023-05-17 at 10 03 29

**requests_count**:

`curl --location --request POST 'http://localhost:8080/v1/histograms' \
--header 'Content-Type: application/json' \
--data-raw '{
    "timeframe": {
        "startTimeUnixNanoSec": 0,
        "endTimeUnixNanoSec": 0
    },
    "filters": [
        {
            "keyValueFilter": {
                "key": "span.kind",
                "operator":"in",
                "value": ["Client"]
            }
        }
    ],
    "interval": 600000,
    "intervalKey": "span.startTimeUnixNano",
    "aggregations": {
        "requests_count": {
            "func": "COUNT"
        }    
    }
}'`

Screenshot 2023-05-17 at 10 03 46

## Which issue(s) this PR fixes:

Fixes #1600

## Checklist

- [X] Tests updated

### CODE REVIEW

**Comment by tvaintrob** [comment]
@Tavh ping me when this is ready i'll review it

File: pkg/model/aggsquery/v1/aggs_query.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+const (
+	PERCENTILES_PARAM AggregationParameter = "percentiles"
+)
+
+type Aggregation struct {
+	Func                  AggregationFunction          `json:"func"`
+	GroupBy               string                       `json:"groupBy,omitempty"`
+	MaxGroups             int                          `json:"maxGroups,omitempty"`
+	Key                   string                       `json:"key,omitempty"`
+	AggregationParameters map[AggregationParameter]any `json:"aggregationParameters,omitempty"`
```
Comment (roybh1) [inline pkg/model/aggsquery/v1/aggs_query.go:58]: Consider changing the name to FunctionParams or AggFunctionParams

File: pkg/model/aggsquery/v1/aggs_query.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+type HistogramsResponse struct {
+	Histograms []Histogram `json:"histograms"`
+}
+
+type Bucket struct {
+	BucketKey any `json:"bucketKey"`
+	Data      any `json:"data"`
+}
+
+type SubBucketsData map[string]any
```
Comment (roybh1) [inline pkg/model/aggsquery/v1/aggs_query.go:76]: Nothing uses this?

Comment (Tavh) [reply pkg/model/aggsquery/v1/aggs_query.go:76]: Yea you're right, it turned out to be kinda pointless.
Removing it

File: pkg/model/aggsquery/v1/aggs_query.go (hunk 1 above)
Comment (Tavh) [reply pkg/model/aggsquery/v1/aggs_query.go:58]: @roybh1 
Yes, probably a clearer name.
Changed it

**Review by roybh1 — APPROVED** [review]
Looks good

**Pull request merged**